Source/hydro/fourth_order.F90

```diff
@@ -56,375 +59,852 @@ subroutine states(idir, &
+                            rho = d2a_lim/d2af(i,j,k)
+                         end if
+
+                         if (rho < 1.0_rt - 1.e-12_rt) then
+                            ! we may need to limit -- these quantities are at cell-centers
+                            d3a_min = min(d3a(i-1,j,k), d3a(i,j,k), d3a(i+1,j,k), d3a(i+2,j,k))
+                            d3a_max = max(d3a(i-1,j,k), d3a(i,j,k), d3a(i+1,j,k), d3a(i+2,j,k))
+
+                            if (C3*max(abs(d3a_min), abs(d3a_max)) <= (d3a_max - d3a_min)) then
+                               ! limit
+                               if (dafm(i,j,k)*dafp(i,j,k) < 0.0_rt) then
+                                  ! Eqs. 29, 30
+                                  ar(i,j,k,n) = a(i,j,k,n) - rho*dafm(i,j,k)  ! note: typo in Eq 29
+                                  al(i+1,j,k,n) = a(i,j,k,n) + rho*dafp(i,j,k)
+                               else if (abs(dafm(i,j,k)) >= 2.0_rt*abs(dafp(i,j,k))) then
+                                  ! Eq. 31
+                                  ar(i,j,k,n) = a(i,j,k,n) - 2.0_rt*(1.0_rt - rho)*dafp(i,j,k) - rho*dafm(i,j,k)
+                               else if (abs(dafp(i,j,k)) >= 2.0_rt*abs(dafm(i,j,k))) then
+                                  ! Eq. 32
+                                  al(i+1,j,k,n) = a(i,j,k,n) + 2.0_rt*(1.0_rt - rho)*dafm(i,j,k) + rho*dafp(i,j,k)
+                               end if
+
+                            end if
+                         end if
+
+                      else
+                         ! if Eqs. 24 or 25 didn't hold we still may need to limit
+                         if (abs(dafm(i,j,k)) >= 2.0_rt*abs(dafp(i,j,k))) then
+                            ar(i,j,k,n) = a(i,j,k,n) - 2.0_rt*dafp(i,j,k)
+                         end if
+                         if (abs(dafp(i,j,k)) >= 2.0_rt*abs(dafm(i,j,k))) then
+                            al(i+1,j,k,n) = a(i,j,k,n) + 2.0_rt*dafm(i,j,k)
+                         end if
+                      end if
+
+                      ! apply flattening
+                      al(i+1,j,k,n) = flatn(i,j,k)*al(i+1,j,k,n) + (ONE - flatn(i,j,k))*a(i,j,k,n)
+                      ar(i,j,k,n) = flatn(i,j,k)*ar(i,j,k,n) + (ONE - flatn(i,j,k))*a(i,j,k,n)
+                   end do
+                end do
+             end do
+
+          end if
+
+       end do  ! component loop
+
+       ! now handle any physical boundaries here by modifying the interface values
+
+       if (lo(1) == domlo(1)) then
 
           do k = lo(3)-dg(3), hi(3)+dg(3)
              do j = lo(2)-dg(2), hi(2)+dg(2)
-                do i = lo(1)-3, hi(1)+3
-                   d2ac(i,j,k) = a(i-1,j,k,n) - 2.0_rt*a(i,j,k,n) + a(i+1,j,k,n)
-                enddo
-             enddo
-          enddo
+
+                ! reset the left state at domlo(1) if needed -- it is outside the domain
+
+                if (physbc_lo(1) == Outflow) then
+                   al(domlo(1),j,k,:) = ar(domlo(1),j,k,:)
```

Comment:
Is this fourth order accurate?

Reply:
I haven't tested the convergence of this, only the reflecting.  reflection was 4th order accurate.  I suspect this is as well, since we are basically telling the Riemann problem that the state is outflow.

Reply:
We've decided to defer this to a future release.